### PR TITLE
improvement(generate:typetests,list): Only write output file when contents differs

### PR DESCRIPTION
**Pull request by tylerbutler**
Adds a utility function to write to files only if the contents you're writing is different from what's already there. This function is used in the type test generation and `flub list` to ensure we only write changes when needed.

This function may be useful for other commands as well, but those changes are left for separate PRs.

I also opportunistically replaced some synchronous function calls with async counterparts.

### CODE REVIEW

File: build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * @param filePath - The path to the file to write.
+ * @param contents - The contents to write to the file.
+ * @returns True if the file was written; false otherwise.
+ */
+export async function writeFileIfContentsDiffers(
+	filePath: PathLike,
+	contents: string,
+): Promise<boolean> {
+	const fileContents = await readFile(filePath, { encoding: "utf8" });
+	if (fileContents !== contents) {
```
Comment (jason-ha) [inline build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts:22]: Any concerns about `\n` versus `\r\n` differences here? (I feel like we have issues with consistency there. Maybe doesn't impact the type test case, but might be a problem for general use.)

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts]: Latest normalizes to LF and clarifies that the function can't be used to output CRLF.

File: build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { type PathLike } from "node:fs";
+import { readFile, writeFile } from "node:fs/promises";
+
+/**
+ * Writes to a file, but first reads the contents to check if it matches the desired content. If it does, the operation
+ * is skipped. Newlines are _always_ normalized to be line-feeds only. This pplies to both the input and the written
```
Comment (alexvy86) [inline build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts:11]:
```suggestion
 * is skipped. Newlines are _always_ normalized to be line-feeds only. This applies to both the input and the written
```

File: build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * @param contents - The contents to write to the file.
+ * @returns True if the file was written; false otherwise.
+ */
+export async function writeFileIfContentsDiffers(
+	filePath: PathLike,
+	contents: string,
+): Promise<boolean> {
+	const fileContents = await readFile(filePath, { encoding: "utf8" });
+	const normalizedFileContents = fileContents.replace(/\r\n/g, "\n");
+	const normalizedNewContents = contents.replace(/\r\n/g, "\n");
```
Comment (alexvy86) [inline build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts:24]: Applying this to the current file contents means that we could inadvertently leave CLRFs in the file if they somehow get there. Would it be better to only apply it to the new content? If thing differ only because of newlines, we probably want to rewrite the file anyway, I think.

Comment (jason-ha) [reply build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts:24]: Actually, not likely that we want to rewrite. Newline difference won't matter to parsing the file for any of our uses. Git settings should be normalizing files. (There are several options for how that works.) The common thing I see with tools that don't respect the current newlines is that files is shown as changed and then when staged git fixes the file up. I see this all the time with the `feeds/public|internal*.txt` files - they show changed and then delta disappear once staged.

File: build-tools/packages/build-cli/src/commands/generate/typetests.ts [hunk 4]
```diff
@@ -190,8 +191,15 @@ declare type MakeUnusedImportErrorsGoAway<T> = TypeOnly<T> | MinimalType<T> | Fu
 
 		mkdirSync(outDir, { recursive: true });
```
Comment (jason-ha) [inline build-tools/packages/build-cli/src/commands/generate/typetests.ts:192]: option: make this async now too.

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/generate/typetests.ts]: Replaced this and rmSync with their async counterparts in latest.

File: build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts (hunk 3 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts:24]: @jason-ha I updated this PR to use the new function for `flub list`, too, so that should help with the feed txt files.

File: build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+	if (normalizedFileContents !== normalizedNewContents) {
+		await writeFile(filePath, normalizedNewContents, "utf8");
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts]: Fixed in latest.

**Comment by CraigMacomber** [comment]
I think the motive I had for suggesting this work (avoid needless file churn as it could invalidate incremental build state) was flawed for multiple reasons:

1. We don't use modification time of inputs vs outputs to do our incremental build
2. If we did use the input vs output build times, tasks which consume other outputs could end up forever dirty due to this change, which is what I was thinking it might fix.

In conclusion: If I was right about how our build worked (and I wasn't) this would be really bad. Since I don't actually know how this part of our build works, someone else should probably make the judgement on if we want to do this.

File: build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * @param filePath - The path to the file to write.
+ * @param contents - The contents to write to the file.
+ * @returns True if the file was written; false otherwise.
+ */
+export async function writeFileIfContentsDiffers(
+	filePath: PathLike,
+	contents: string,
+): Promise<boolean> {
+	const fileContents = await readFile(filePath, { encoding: "utf8" });
```
Comment (jason-ha) [inline build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts:24]: What happens if the file doesn't exist yet? I would guess it throws, but in that case we want an explicit write.

File: build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ * Writes to a file, but first reads the contents to check if it matches the desired content. If it does, the operation
+ * is skipped. Newlines are _always_ normalized to be line-feeds only for comparison. This effectively means that
+ * newline differences are ignored when comparing the contents. However, if the file is written, then it will be written
+ * as provided. For example, if `contents` is using CRLF newlines and the file is written, it will be written with CRLF
+ * newlines.
+ *
+ * @param filePath - The path to the file to write.
+ * @param contents - The contents to write to the file.
+ * @returns True if the file was written; false otherwise.
+ */
```
Comment (jason-ha) [inline build-tools/packages/build-cli/src/library/writeFileIfContentsDiffers.ts:19]: Can we add a note about performance improvements?
This helper fully reads and closes a file before writing it (if needed).
In the common case that writing is needed, it will be more efficient to open the file for read+write to start and keep it open until any writing is dealt with.

**Review by jason-ha — COMMENTED** [review]
I still think it is a better idea to understand the root trouble with .done tracking before making things more complicated.
